oauth2_provider/views/oidc.py

```diff
@@ -347,21 +349,83 @@ def form_valid(self, form):
 
-            if not prompt or form.cleaned_data.get("allow"):
-                return self.do_logout(application, redirect_uri, state, token_user)
+            if not self.must_prompt(token_user) or form.cleaned_data.get("allow"):
+                return self.do_logout(application, post_logout_redirect_uri, state, token_user)
             else:
                 raise LogoutDenied()
 
         except OIDCError as error:
             return self.error_response(error)
 
+    def validate_logout_request(self, id_token_hint, client_id, post_logout_redirect_uri):
+        """
+        Validate an OIDC RP-Initiated Logout Request.
+        `(application, token_user)` is returned.
+
+        If it is set, `application` is the Application that is requesting the logout.
+        `token_user` is the id_token user, which will used to revoke the tokens if found.
+
+        The `id_token_hint` will be validated if given. If both `client_id` and `id_token_hint` are given they
+        will be validated against each other.
+        """
+
+        id_token = None
+        if id_token_hint:
+            # Only basic validation has been done on the IDToken at this point.
+            id_token, claims = _load_id_token(id_token_hint)
+
+            if not id_token or not _validate_claims(self.request, claims):
+                raise InvalidIDTokenError()
+
+            # If both id_token_hint and client_id are given it must be verified that they match.
+            if client_id:
+                if id_token.application.client_id != client_id:
+                    raise ClientIdMissmatch()
+
+        application = None
+        # Determine the application that is requesting the logout.
+        if client_id:
+            application = get_application_model().objects.get(client_id=client_id)
+        elif id_token:
+            application = id_token.application
+
+        # Validate `post_logout_redirect_uri`
+        if post_logout_redirect_uri:
+            if not application:
+                raise InvalidOIDCClientError()
+            scheme = urlparse(post_logout_redirect_uri)[0]
+            if not scheme:
+                raise InvalidOIDCRedirectURIError("A Scheme is required for the redirect URI.")
+            if oauth2_settings.OIDC_RP_INITIATED_LOGOUT_STRICT_REDIRECT_URIS and (
+                scheme == "http" and application.client_type != "confidential"
+            ):
+                raise InvalidOIDCRedirectURIError("http is only allowed with confidential clients.")
+            if scheme not in application.get_allowed_schemes():
+                raise InvalidOIDCRedirectURIError(f'Redirect to scheme "{scheme}" is not permitted.')
+            if not application.post_logout_redirect_uri_allowed(post_logout_redirect_uri):
+                raise InvalidOIDCRedirectURIError("This client does not have this redirect uri registered.")
+
+        return application, id_token.user if id_token else None
+
+    def must_prompt(self, token_user):
```

Comment:
The reason I wanted to override must_prompt is for the case where we are given a id_token_hint that matches a user that already logged out (because he already logged out through another RP).

In this case DOT prompt the user event if the actual logout does nothing (no session to clean, no token to revoke).
 
We felt that it was not useful to prompt for logout because we would not do anything afterward, and so we had to override it.

I'm still unclear it's supposed to work in the specs :
- there's nothing explicitely told about prompting if there's no ongoing session
- and we can read [here](https://openid.net/specs/openid-connect-rpinitiated-1_0.html#ValidationAndErrorHandling) that ``RP-Initiated Logout Requests are intended to be idempotent`` so prompting only if the user already logged out form another RP before does seems counter intuitive.

Reply:
What does the specification for the prompt endpoint say to do if there is no session for the user? In this case, Given the constraint of an idempotent end sessions endpoint, I would expect the endpoint to always behave the same give a set of inputs.  ie) is there was no sessions it would still prompt. Then the prompt code, if there was no sessions, could just redirect the user.  An approach of deferring to session check to the logout prompt would allow redirects to federated IDPs logout prompts... While not a use case I think we will support in DOT it is not uncommon that an OIDC IDP might be federated with some other enterprise authentication system.

Reply:
I think my brain crossed wires with pure function and idempotent for a second.  I think both approached are fine. Personally  I think it would be cleaner to keep the proscribed OIDC flow, and do the redirect in the prompt code.

Reply:
The only mention I found in the specs about logged out users is in the `Validation and Error Handling` section, just after the `idempotent` part : it's not an error to try login out a user that's not currently logged in.  

I also feel we should keep the current implementation the way it is, and just let users of DOT override it the way they want.